xapian-core/cluster/cluster.cc

```diff
@@ -324,91 +349,189 @@ Point
+{
+}
+
+doccount
+ClusterSet::Internal::size() const
+{
+    return clusters.size();
+}
+
 doccount
 ClusterSet::size() const
 {
     LOGCALL(API, doccount, "ClusterSet::size", NO_ARGS);
-    return clusters.size();
+    return internal->size();
 }
 
 void
-ClusterSet::add_cluster(Cluster &c)
+ClusterSet::Internal::add_cluster(const Cluster &c)
 {
-    LOGCALL_VOID(API, "ClusterSet::add_cluster", c);
     clusters.push_back(c);
 }
 
+void
+ClusterSet::add_cluster(const Cluster &c)
+{
+    LOGCALL_VOID(API, "ClusterSet::add_cluster", c);
+    internal->add_cluster(c);
+}
+
 Cluster&
-ClusterSet::operator[](doccount i)
+ClusterSet::Internal::get_cluster(doccount i)
 {
     return clusters[i];
 }
 
+Cluster&
+ClusterSet::operator[](doccount i)
+{
+    return internal->get_cluster(i);
+}
+
+void
+ClusterSet::Internal::add_to_cluster(const Point &x, unsigned int i)
+{
+    clusters[i].add_point(x);
+}
+
 void
 ClusterSet::add_to_cluster(const Point &x, unsigned int i)
 {
    LOGCALL_VOID(API, "ClusterSet::add_to_cluster", x | i);
-   clusters[i].add_point(x);
+   internal->add_to_cluster(x, i);
 }
 
 void
-ClusterSet::recalculate_centroids()
+ClusterSet::Internal::recalculate_centroids()
 {
-    LOGCALL_VOID(API, "ClusterSet::recalculate_centroids", NO_ARGS);
     for (vector<Cluster>::iterator it = clusters.begin(); it != clusters.end(); ++it)
 	(*it).recalculate();
 }
 
+void
+ClusterSet::recalculate_centroids()
+{
+    LOGCALL_VOID(API, "ClusterSet::recalculate_centroids", NO_ARGS);
+    internal->recalculate_centroids();
+}
+
+void
+ClusterSet::clear_clusters()
+{
+    LOGCALL_VOID(API, "ClusterSet::clear_clusters", NO_ARGS);
+    internal->clear_clusters();
+}
+
+void
+ClusterSet::Internal::clear_clusters()
+{
+    for (vector<Cluster>::iterator it = clusters.begin(); it !=clusters.end(); ++it)
+	(*it).clear();
+}
+
 doccount
 Cluster::size() const
 {
     LOGCALL(API, doccount, "Cluster::size", NO_ARGS);
+    return internal->size();
+}
+
+doccount
+Cluster::Internal::size() const
+{
     return (cluster_docs.size());
 }
 
 void
 Cluster::add_point(const Point &doc)
 {
     LOGCALL_VOID(API, "Cluster::add_point", doc);
+    internal->add_point(doc);
+}
+
+void
+Cluster::Internal::add_point(const Point &doc)
+{
     cluster_docs.push_back(doc);
 }
 
 void
 Cluster::clear()
 {
     LOGCALL_VOID(API, "Cluster::clear", NO_ARGS);
+    internal->clear();
+}
+
+void
+Cluster::Internal::clear()
+{
     cluster_docs.clear();
 }
 
 Centroid
 Cluster::get_centroid() const
 {
     LOGCALL(API, Centroid, "Cluster::get_centroid", NO_ARGS);
+    return internal->get_centroid();
+}
+
+Centroid
+Cluster::Internal::get_centroid() const
+{
     return centroid;
 }
 
 void
 Cluster::set_centroid(const Centroid centroid_)
 {
     LOGCALL_VOID(API, "Cluster::set_centroid", centroid_);
-    centroid = centroid_;
+    internal->set_centroid(centroid_);
 }
 
 void
-ClusterSet::clear_clusters()
+Cluster::Internal::set_centroid(const Centroid centroid_)
```

Comment:
Reference?

Please can you check for any further instances too.

If they shouldn't be references, then the `const` is odd and should probably be removed.